File: 2024/02/main.clj
(ns lv.kristapsbe.advent)

(defn get-input [file-path]
    (let [lines (slurp file-path)
          data (mapv #(clojure.string/split % #"\s+") (clojure.string/split-lines lines))
        ] data))

(defn cast-values [f coll]
    (mapv (fn [s] (if (vector? s) (cast-values f s) (f s))) coll))

(defn calculate-adj [e coll v]
    (if (some? coll) (calculate-adj (first coll) (next coll) (conj v (- (first coll) e))) v))

(defn is-safe [coll]
    (if (= (abs (apply + (map #(if (and (> (abs %) 0) (< (abs %) 4)) (max (min % 1) -1) 0) coll))) (count coll)) 1 0))

(defn vec-remove [pos coll]
    (into (subvec coll 0 pos) (subvec coll (inc pos))))

(defn add-part-two-options [ct coll v]
    (let [next-ct (inc ct)]
        (if (= ct (count coll)) v (add-part-two-options next-ct coll (conj v (vec-remove ct coll))))))

(defn any-safe [row]
    (apply max (map #(is-safe (calculate-adj (first %) (next %) [])) row)))

(def data (cast-values Integer/parseInt (get-input "input.txt")))
(def adj (map #(calculate-adj (first %) (next %) []) data))
(println (apply + (map is-safe adj))) ; part one
(def part-two-data (map #(add-part-two-options 0 % [%]) data))
(println (apply + (map any-safe part-two-data))) ; part two

File: 2024/02/main.py
reports = [[int(e) for e in l.split()] for l in open("input.txt", "r").readlines()]

max_min = range(1, 4)

adj = [list(map(lambda e: e[0]-e[1], zip(r, r[1:]))) for r in reports]
def is_safe(a):
    return 1 if abs(sum([max(-1, min(1, e if abs(e) in max_min else 0)) for e in a])) == len(a) else 0

print(sum([is_safe(a) for a in adj]))

all_reports = [[r[:i]+r[i+1:] for i in range(len(r))]+[r] for r in reports]
adj = [[list(map(lambda e: e[0]-e[1], zip(r, r[1:]))) for r in ar] for ar in all_reports]
print(sum([max([is_safe(e) for e in a]) for a in adj]))
